React Renderer: refactor the code, clean it up and make it more generic

mikehearn opened the pull request:
This series of commits is functionally nearly identical to the current code (save for a bug fix to the sandbox that nobody uses yet). However it improves the code in these ways:

1. It's smaller and simpler now.
2. The `utils` sub-package contains classes that could be integrated into Micronaut Core. They aren't React, Graal or Views specific.
3. The `truffle` sub-package contains classes that could be re-used for a more generic Truffle integration.
4. More objects are overridable via factories, and beans are now properly namespaced using qualifiers.
5. The docs for how to use hot-reload have been improved.

I hope that in future the generic code can find a new home in other modules.

mikehearn committed:
ReactJS: Update README.

mikehearn committed:
ReactJS: Refactor file reload handler out of JSContextPool, mark some fields private.

mikehearn committed:
ReactJS: Improve the docs for how to enable hot file reload.

mikehearn committed:
ReactJS: Refactor JSContextPool into BeanPool.

This is a new generic utility useful for not only other Truffle languages, but Micronaut in general.

Improved the documentation, efficiency and usability of the class along the way.

mikehearn committed:
ReactJS: Move BeanPool into a sub-package.

mikehearn committed:
ReactJS: Move OutputStreamToSLF4J into the utils sub-package.

This class is generally useful and could be a part of Micronaut (or really, SLF4J itself).

mikehearn committed:
ReactJS: Move the JUL2SLF4J adapter into the utils sub-package.

This class is generally useful and could be a part of Micronaut (or really, SLF4J itself).

mikehearn committed:
ReactJS: Move and rename the proxy adapter class.

mikehearn committed:
ReactJS: Make BeanPool optionally close beans when they are cleared or obsoleted.

mikehearn committed:
ReactJS: Fix an obsolete JavaDoc.

mikehearn committed:
ReactJS: Refactoring to move Engine creation to a bean factory.

This makes the code a bit more generic.

mikehearn committed:
ReactJS: Bugfix to the sandbox policy.

List/map access wasn't previously allowed, so in sandbox mode lists and maps wouldn't be exposed to JS and would appear empty.

mikehearn committed:
ReactJS: Rename some classes to make it clear they are React specific and not JS specific.

mikehearn committed:
ReactJS: Refactor more code into the bean factory.

mikehearn committed:
ReactJS: Use qualifiers to avoid conflicts with other Micronaut code that uses Truffle classes.

mikehearn committed:
ReactJS: Refactor away the CompiledReactJSBundle class.

It's no longer really needed.

mikehearn committed:
ReactJS: Use c'tor injection.

mikehearn committed:
ReactJS: Extract source builder method.

mikehearn committed:
ReactJS: Simplify source builder logic.

mikehearn committed:
ReactJS: Unify more source builder logic and simplify again.

mikehearn committed:
ReactJS: checkstyle fixes

mikehearn committed:
ReactJS: javadoc++

mikehearn committed:
ReactJS: work around a Truffle bug by going back to reading the bundle resources ourselves.

sdelamo commented on views-react/src/main/java/io/micronaut/views/react/util/JavaUtilLoggingToSLF4J.java:
```diff
@@ -13,12 +13,10 @@
  * limitations under the License.
  */
-package io.micronaut.views.react;
+package io.micronaut.views.react.util;
```
this class should be in [micronaut-logging](https://github.com/micronaut-projects/micronaut-logging)

sdelamo commented on views-react/src/main/java/io/micronaut/views/react/util/OutputStreamToSLF4J.java:
```diff
@@ -30,7 +30,7 @@
  */
 @Internal
-final class OutputStreamToSLF4J extends OutputStream {
+public final class OutputStreamToSLF4J extends OutputStream {
```
this class should in [Micronaut Logging](https://github.com/micronaut-projects/micronaut-logging)

mikehearn replied:
OK, I'll open a PR against that module to add these classes after this PR is merged. This PR unblocks such moves by refactoring the existing code.

sdelamo commented on views-react/src/main/java/io/micronaut/views/react/ReactJSBeanFactory.java:
```diff
@@ -0,0 +1,136 @@
+
+    // We cache the Source objects because they are expensive to create, but, we don't want them
+    // to be singleton beans so we can recreate them on file change.
+    private Source serverBundle;  // L(this)
```
can we use a different scope, e.g. `@Refreshable` instead of this custom caching?

mikehearn replied:
Maybe, I'm not sure. It's triggered by a file watch event and has to be integrated with the BeanPool.

sdelamo commented on views-react/src/main/java/io/micronaut/views/react/ReactJSContext.java:
```diff
@@ -15,29 +15,27 @@
  */
 @Internal
-class JSContext implements AutoCloseable {
+@Bean
```
should we `@Singleton` instead of `@Bean`?

mikehearn replied:
No, it's not a singleton object. There needs to be multiple of them to enable concurrency.

sdelamo commented on views-react/src/main/java/io/micronaut/views/react/util/BeanPool.java:
```diff
@@ -0,0 +1,180 @@
+ * @param <T> The type of the bean that being pooled.
+ */
+@Internal
+public class BeanPool<T> {
```
do we need in Micronaut Views? Or is this adding an extra layer of complexity? Should this API be in core?

mikehearn replied:
It's necessary complexity but it could be in Micronaut Core, as the PR says.

Let's get this PR merged first, then code can be moved out into other modules.

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/ReactJSBeanFactory.java:
```diff
@@ -0,0 +1,136 @@
+ */
+@Factory
+@Internal
+class ReactJSBeanFactory {
```
can probably be `final`

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/ReactJSBeanFactory.java:
```diff
@@ -0,0 +1,136 @@
+     * allowing sandboxed JS to extend or implement Java types.
+     */
+    @Singleton
+    @Named("react")
```
extract these names into constants

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/ReactJSContext.java:
```diff
@@ -46,89 +44,69 @@ class JSContext implements AutoCloseable {
-        this.compiledJS = compiledJS;
+    ReactJSContext(ReactViewsRendererConfiguration configuration,
+                   @Named("react") Engine engine,
+                   @Named("react") HostAccess hostAccess,
```
reuse constants here

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/ReactJSContext.java:
```diff
@@ -15,29 +15,27 @@
-import org.graalvm.polyglot.Source;
-import org.graalvm.polyglot.Value;
+import jakarta.inject.Named;
+import org.graalvm.polyglot.*;
```
don't use wildcard imports

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/util/BeanPool.java:
```diff
@@ -0,0 +1,180 @@
+ */
+@Internal
+public class BeanPool<T> {
+    // TODO: Use @Scheduled to occasionally clear out beans that weren't accessed for a while to recover from traffic spikes.
```
address this TODO

mikehearn replied:
Shall we do it in a separate PR? This one is trying to be a functionality no-op as far as possible.

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/util/BeanPool.java:
```diff
@@ -0,0 +1,180 @@
+    private final Supplier<T> factory;
+
+    // Synchronized on 'this'.
+    private final LinkedList<SoftReference<PoolEntry>> pool = new LinkedList<>();
```
Since this required cleanup you may be better off using the `ConcurrentLinkedHashMap` implementation we have in `io.micronaut.core.util.clhm` with a size limit to implement recovery

mikehearn replied:
Thanks for the tip. I looked at using it but that would require making the entire BeanPool class lock free which would be easy to subtly screw up. Given that the expensive part is rendering the HTML the BeanPool lock is unlikely to be a bottleneck and hasn't been so in any of the profiling I've done. I'd worry that by trying to use this class I'd accidentally introduce race conditions.

mikehearn replied:
To clarify, having a size limit is good (presumably with a semaphore to pause threads if too many contexts are checked out), but it's probably better to use a regular LinkedHashMap and lock here as that way the logic in checkOut can be straightforward and take into account file reloads, age, etc. This feels like a separate PR's worth of logic.

mikehearn committed:
ReactJS: Extract the qualifier to a constant, mark the ReactJSBeanFactory as final.

mikehearn commented:
Can this be merged?

dstepanov commented on views-react/src/main/java/io/micronaut/views/react/ReactJSBeanFactory.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Bean
+    @Named(REACT_QUALIFIER)
+    synchronized Source serverBundle(ResourceResolver resolver, ReactViewsRendererConfiguration reactConfiguration) throws IOException {
```
Why do you need `synchronized`? If the bean is a singleton it should be created only once

mikehearn replied:
They aren't singletons, they can be invalidated by file changes on disk and re-created in response to that event (see the code in `maybeReloadServerBundle`). When that happens the bean pool is flushed and the contexts are recreated.

dstepanov commented on views-react/src/main/java/io/micronaut/views/react/util/BeanPool.java:
```diff
@@ -0,0 +1,180 @@
+     * @return a wrapper for the entry that you should pass to {@link #checkIn(Handle)} when you're
+     * done with it to put it back into the pool.
+     */
+    public synchronized Handle<T> checkOut() {
```
I would prefer to use a lambda function to prevent accident missed close:

```
beanPool.<ResultType>useContext(ctx -> {
// use context
return result;
});
```

mikehearn replied:
Done

mikehearn committed:
ReactJS: Add a lambda helper to BeanPool.

mikehearn commented:
@dstepanov PTAL

dstepanov commented:
I didn't check the PR more closely before, so now I have a better look.
Instead of doing in-code comments, I will try to summarise it here:
- `@Named(REACT_QUALIFIER)` - it would be better to have a custom qualifier annotation
- Not sure why do you need to create `NamedSourceQualifier`
- `Context.newBuilder()` options might be better to be provided as a configuration (if it makes sense) `Context.Builder` might be represented as `@ConfigurationBuilder`. The same might be applicable for `Engine.newBuilder`
- Having some synchronized state in a factory is very strange and should be grouped with the actual state
- Essentially, it looks like the bean pool should be created from some configuration state that will hold all the sources, and if the event changes the configuration, the pool should be purged, and the configuration should be recreated with the new values. This logic of the clearing the pool shouldn't be part of `ReactViewsRenderer`

IMHO, the best would be to have `BeanPool` (as an interface, preferably) and a `BeanPoolInstanceProvider` with a callback that the pool will use to purge the bean pool.

dstepanov commented:
Sometimes, we would make builders as beans which will allow users to alter the configuration

mikehearn commented:
Being able to control the context via config etc is a good idea, but it feels like a separate PR. I'm  trying to keep this one a functionality no-op, just a refactoring. It's already got quite large.

The BeanPool is purged when source files on disk change not when configuration changes, so I didn't fully understand that suggestion, apologies. The existing refresh mechanisms don't seem applicable here.

When you say group the synchronized methods with the state, could you elaborate what you have in mind? The style linter doesn't allow fields to be moved next to the methods that manipulate them. Do you mean move the file change listener into the factory? The issue is that the things in the BeanPool are vended by the Factory, so the factory has to be told to re-create the beans in the same way as initially, but it's the owner of the pool that has to clear them.

dstepanov commented:
I’m talking about the design. The configuration should be affected by the listener, and the bean pool should be affected by the configuration. Now it’s a bit messy.

mikehearn commented:
Which configuration do you mean? The bean pool is only affected by a change in the checksum of the files on disk, which isn't reflected in any configuration bean. This is for hot reload scenarios.

dstepanov commented:
I call all the files that affect the context instance as configuration.

mikehearn committed:
ReactJS: Fix BeanPool javadoc.

mikehearn committed:
ReactJS: More refactorings.

1. ContextPoolManager now encapsulates source loading and reloading.
2. Encapsulated @Named("react") behind a custom annotation.
3. The named qualifier resolver is no longer needed.

mikehearn commented:
@dstepanov How does it look now?

dstepanov committed:
ReactJS: Even more refactorings.

dstepanov commented:
@mikehearn Please check https://github.com/mikehearn/micronaut-views/pull/1

dstepanov commented:
@mikehearn Can you please resolve conflicts and make sure the build is passing?

mikehearn committed:
ReactJS: Fix a bean conflict introduced by the merge and some spotless errors.

mikehearn commented:
Looks like the Java 21 build hit a Gradle bug?!

mikehearn committed:
React: bump GraalJS and turn on sandboxing tests.

This version fixes a JVM crash that occurs with the latest GraalVM, and removes the need for the prior test disables related to sandboxing.

The JSON object ordering isn't stable across versions so fix the assumption that it is.

mikehearn committed:
React: reactivate on CI now JVM crash is avoided.

mikehearn committed:
Merge branch '5.6.x' into reactjs/make-it-more-generic
